actors/sac.py

```diff
@@ -0,0 +1,619 @@
+# Copyright 2022 AI Redefined Inc. <dev+cogment@ai-r.com>
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import time
+from typing import Tuple
+
+import cogment
+import numpy as np
+import torch
+from gym.spaces import Box, utils
+from torch.distributions.normal import Normal
+
+from cogment_verse import Model, TorchReplayBuffer, TorchReplayBufferSample
+from cogment_verse.run.run_session import RunSession
+from cogment_verse.run.sample_producer_worker import SampleProducerSession
+from cogment_verse.specs import PLAYER_ACTOR_CLASS, AgentConfig, EnvironmentConfig, EnvironmentSpecs, cog_settings
+
+# torch.multiprocessing.set_sharing_strategy("file_system")
+torch.set_num_threads(1)
+
+log = logging.getLogger(__name__)
+LOG_STD_MAX = 2
+LOG_STD_MIN = -20
+
+# pylint: disable=not-callable
+class PolicyNetwork(torch.nn.Module):
+    """Gaussian policy network where action is modeled by Gaussian distribution"""
+
+    def __init__(self, num_input: int, num_output: int, num_hidden: int) -> None:
+        super().__init__()
+        self.input = torch.nn.Linear(num_input, num_hidden)
+        self.fully_connected = torch.nn.Linear(num_hidden, num_hidden)
+        self.mean = torch.nn.Linear(num_hidden, num_output)
+        self.log_std = torch.nn.Linear(num_hidden, num_output)
+        # self.log_std = torch.nn.Parameter(torch.zeros(1, num_output))
+
+    def forward(self, x: torch.Tensor) -> Tuple[Normal, torch.Tensor]:
+        # Input layer
+        x = self.input(x)
+        x = torch.nn.functional.relu(x)
+
+        # Hidden layer
+        x = self.fully_connected(x)
+        x = torch.nn.functional.relu(x)
+
+        # Output layer
+        mean = self.mean(x)
+        log_std = self.log_std(x)
+
+        return mean, log_std
+
+
+class ValueNetwork(torch.nn.Module):
+    """Value network that quantifies the quality of an action given a state."""
+
+    def __init__(self, num_input: int, num_hidden: int):
+        super().__init__()
+        # Value network
+        self.input = torch.nn.Linear(num_input, num_hidden)
+        self.fully_connected = torch.nn.Linear(num_hidden, num_hidden)
+        self.output = torch.nn.Linear(num_hidden, 1)
+
+    def forward(self, observation_action: torch.Tensor) -> torch.Tensor:
+        # Value network
+        x_v = self.input(observation_action)
+        x_v = torch.nn.functional.relu(x_v)
```

Comment:
to be consistent with rest of our code base, maybe you could import torch.nn.functional as F and just `F.relu(self.input(observation_action))`. It's also preferred if you just use `self.linear1`, `self.linear2` etc.. instead of `self.input`, `self.fully_connected` etc?

Reply:
Done